### PR TITLE
MGMT-18169: Set conditions for device heartbeats

### PR DESCRIPTION
**Pull request by avishayt**
1. Adds device.spec.settings where the user can set warning and error-level thresholds for device liveness.
2. When the Agent calls device.UpdateStatus(), the service:
   - Sets device.status.lastSeen with the current time. This property was previously called updatedAt but we decided to change it (API breakage alert).
   - Sets an internal DB property of the time at which this device's liveness will expire. This is used for efficiently finding expired devices without iterating over all of them.
   - Unsets any liveness warning/error condition.
3. A new task periodically fetches any expired devices and sets the warning/error conditions as appropriate.

### CODE REVIEW

**Review by fzdarsky — COMMENTED** [review]
I'm wondering whether there's a simpler way that allows us to avoid adding more conditions and running a task: API server sets the `lastSeen` property of `deviceStatus`, in the model we pull this and the `disconnectedThreshold` into separate columns, where `lastSeen + disconnectedThreshold < now()` can be computed at any time. Means, we can lazily compute it when getting a device record or when we need to compute the statuses.

**Comment by avishayt** [comment]
> I'm wondering whether there's a simpler way that allows us to avoid adding more conditions and running a task: API server sets the `lastSeen` property of `deviceStatus`, in the model we pull this and the `disconnectedThreshold` into separate columns, where `lastSeen + disconnectedThreshold < now()` can be computed at any time. Means, we can lazily compute it when getting a device record or when we need to compute the statuses.

I dislike piggybacking on get/list calls to compute fields.  The UI and others make these API calls rather often and should be as efficient as possible - everything computed in advance and simply returned to the user.

File: api/v1alpha1/openapi.yaml [hunk 1]
```diff
@@ -2175,6 +2176,17 @@ components:
                 type: string
                 pattern: '^[a-zA-Z0-9:\-_.\\*]*$' # * plus https://github.com/systemd/systemd/blob/9cebda59e818cdb89dc1e53ab5bb51b91b3dc3ff/src/basic/unit-name.c#L42
                 maxLength: 256
+        settings:
+          $ref: '#/components/schemas/DeviceSettings'
+    DeviceSettings:
+      type: object
+      properties:
+        heartbeatWarningTime:
```
[the remark this answers is not in the document]
Comment (avishayt) [reply api/v1alpha1/openapi.yaml:2184]: I thought it was a health issue and cause for concern/alarm.  I made it configurable to handle cases where the user expects the agent not to check in often.

File: api/v1alpha1/openapi.yaml [hunk 2]
```diff
@@ -2175,6 +2176,17 @@ components:
                 type: string
                 pattern: '^[a-zA-Z0-9:\-_.\\*]*$' # * plus https://github.com/systemd/systemd/blob/9cebda59e818cdb89dc1e53ab5bb51b91b3dc3ff/src/basic/unit-name.c#L42
                 maxLength: 256
+        settings:
```
[the remark this answers is not in the document]
Comment (avishayt) [reply api/v1alpha1/openapi.yaml:2179]: Right, settings around how the device should be treated, but not what's written to disk.  How often to send a heartbeat, timeouts on heartbeats, maybe things like log level?

[the remark this answers is not in the document]
Comment (mangelajo) [reply api/v1alpha1/openapi.yaml:2179]: in the RFE for the API we were proposing a "monitoring" section where we would add containers an systemd. Should we also add those settings (as it's about monitoring)

https://github.com/flightctl/flightctl/blob/main/docs/enhancements/fep-000-api-device-fleet.md#devicesname-user-facing-method

How about: ? 
```
monitoring:
     containers:
     systemd:
     heartbeat:
         warningTime:
         errorTime:
```

[the remark this answers is not in the document]
Comment (avishayt) [reply api/v1alpha1/openapi.yaml:2179]: I think that makes sense

[the remark this answers is not in the document]
Comment (hexfusion) [reply api/v1alpha1/openapi.yaml:2179]: Monitoring is a device.spec. are we staying that we are adding a monitoring status? Both of these configs are service level right?

File: api/v1alpha1/openapi.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (hexfusion) [reply api/v1alpha1/openapi.yaml:2184]: Would depend a lot on ENV. Factory floor would potentially consider this a stop the world event.

File: api/v1alpha1/openapi.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (fzdarsky) [reply api/v1alpha1/openapi.yaml:2179]: On the container/systemd status: Keep in mind we added these in the early days just to illustrate the live view we’re aiming at, knowing they’ll likely have to look differently. With the application status we’re modelling, at least the container and possibly both may become obsolete.

[the remark this answers is not in the document]
Comment (fzdarsky) [reply api/v1alpha1/openapi.yaml:2179]: On possible alternative names / structure for “settings”: The status update interval is an agent concern, therefore part of the agent config file, not the device.spec. And while containers/systemd units are monitored by the agent, the status updates / heartbeats are monitored by the service. Seems weird to put user intent on different objects under the same heading.

File: api/v1alpha1/openapi.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (fzdarsky) [reply api/v1alpha1/openapi.yaml:2184]: I agree, but we’re talking about status updates sent at the rate of 1 per X minutes, hours, or days. On a factory floor you’d notice production stopping much faster and you’d use other mechanisms to detect connectivity issues in milliseconds to seconds.

**Comment by fzdarsky** [comment]
> I dislike piggybacking on get/list calls to compute fields. The UI and others make these API calls rather often and should be as efficient as possible - everything computed in advance and simply returned to the user.

Maybe instead of “piggybacking” (which sounds unrelated) call it “late materialization”? True, per API call overhead is slightly higher, the question is where is the scalability
bottleneck, CPU or network I/O? For the latter, asks add to the load (and if we want to produce useful alarms, we probably want to run them frequently?) while generating on demand is free.

**Comment by avishayt** [comment]
> > I dislike piggybacking on get/list calls to compute fields. The UI and others make these API calls rather often and should be as efficient as possible - everything computed in advance and simply returned to the user.
> 
> Maybe instead of “piggybacking” (which sounds unrelated) call it “late materialization”? True, per API call overhead is slightly higher, the question is where is the scalability bottleneck, CPU or network I/O? For the latter, asks add to the load (and if we want to produce useful alarms, we probably want to run them frequently?) while generating on demand is free.

How would we produce an event/alarm?  We would need to wait until someone GETs the device and then write to the DB during the GET call to remember that we produced the event/alarm.

File: api/v1alpha1/openapi.yaml [hunk 3]
```diff
@@ -1797,8 +1797,9 @@ components:
     DeviceStatus:
       type: object
       properties:
-        updatedAt:
+        lastSeenAt:
```
Comment (mangelajo) [inline api/v1alpha1/openapi.yaml:1800]: can we update in this PR: https://github.com/flightctl/flightctl/blob/main/docs/enhancements/fep-000-api-device-fleet.md

https://github.com/flightctl/flightctl/blob/main/docs/enhancements/fep-000-api-device-fleet.md#devicesname-user-facing-method on the yaml example

and then

https://github.com/flightctl/flightctl/blob/main/docs/enhancements/fep-000-api-device-fleet.md#devicesnamestatus-device-side

Where it talks about the heartbeat mechanism?

File: api/v1alpha1/openapi.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mangelajo) [reply api/v1alpha1/openapi.yaml:2179]: > Monitoring is a device.spec. are we staying that we are adding a monitoring status? Both of these configs are service level right?

Not status, just configuration for monitoring in the device spec.

[the remark this answers is not in the document]
Comment (mangelajo) [reply api/v1alpha1/openapi.yaml:2179]: > On possible alternative names / structure for “settings”: The status update interval is an agent concern, therefore part of the agent config file, not the device.spec. And while containers/systemd units are monitored by the agent, the status updates / heartbeats are monitored by the service. Seems weird to put user intent on different objects under the same heading. As Avishay said, this is more about “how the device should be treated” or managed by the service, so maybe “managementSettings” or “managementPolicies” or something in that direction?

I think this should be completely dynamic and not on the agent config.

It should be provided by the server:

1) hey agent, this is how often I want you to pick up config
2) hey agent, this is how often I want you to report status (i.e. 1/2 of the timeout time)

then we need to declare the desired timeout somewhere, I think the device is fine. But also. We must keep in mind that this is likely to be dynamic at some point. The server detects the load of the system, and throttles the reconciliation loop by asking the devices to go request slower....  or a device has to be updated, and it's preemptively asked to start polling/reporting more often, then brought back to the normal schedule.

File: api/v1alpha1/openapi.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mangelajo) [reply api/v1alpha1/openapi.yaml:2184]: dynamic! :)

File: api/v1alpha1/openapi.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (fzdarsky) [reply api/v1alpha1/openapi.yaml:2179]: Ok, I see that use case. Then it sounds like we want to structure that under `device.spec.agent` or `device.spec.management`, though?

[the remark this answers is not in the document]
Comment (avishayt) [reply api/v1alpha1/openapi.yaml:2179]: Naming is hard, that's why I just said "settings" :)

**Pull request closed without merging**